### PR TITLE
CUDA: reduce job switch time to sub 10 milliseconds

### PR DESCRIPTION
**Pull request by jean-m-cyr**
Reduces Nvidia job switch time from up to 100ms to sub 10ms.
    
    Resulting in:
    
        - Higher and steadier hash rate.
        - Fewer stale shares.
        - Fully asynchronous stratum and miner threads
    
    Solid for 6 hours on 4 GPU (1060) Ubuntu system
    
    Major rework. Please test.

### CODE REVIEW

**Comment by jean-m-cyr** [comment]
Hash rate steady at 81mh/s, instead of previous fluctuating around 79mh/s.

File: libethash-cuda/CUDAMiner.cpp [hunk 1]
```diff
@@ -509,17 +492,33 @@ void CUDAMiner::search(
...
+						farm.submitProof(Solution{nonces[i], r.mixHash, w.header, w.seed, w.boundary, w.job, w.job_len, m_abort});
+					else
+					{
+						farm.failedSolution();
+						cwarn << "GPU gave incorrect result!";
+					}
+				}
+			}
+			addHashCount(batch_size);
+			if (m_abort || shouldStop())
```
Comment (chfast) [inline libethash-cuda/CUDAMiner.cpp]: The `shouldStop()` also has some kind of "abort" flag. I'm not sure we need both.

Comment (jean-m-cyr) [reply libethash-cuda/CUDAMiner.cpp]: Yeah, that's a little confusing. That should stop is the Farm's shouldStop, not the miner's. It's only used for terminating operations.

**Comment by kkkrackpot** [comment]
@jean-m-cyr I tried `switch` branch from your repo -- the hashrate looks more stable than with my current `ethminer` (I use a rather old one, not current git master).

**Comment by ddobreff** [comment]
Tested for a few hours, looks good to me.

**Comment by jean-m-cyr** [comment]
12 how run.
snap3

**Comment by AndreaLanfranchi** [comment]
Nice job.
Starting a test round right now.

**Comment by chfast** [comment]
Released in v0.13 RC9: https://github.com/ethereum-mining/ethminer/releases/tag/v0.13.0rc9.

**Comment by AndreaLanfranchi** [comment]
@jean-m-cyr and @chfast 
Well ... knew you guys where good. The confirmation has arrived with rc7.

Just ran a 12 hours batch with my rigs and here what ethermine depicts:
2018-01-23 ethermine

Redline marks my start with rc7

- Orange line steadily converges to the green line
- Stale share ratio sticks around 6 to 7.5% max
- Despite the fact MY green line is a little bumpy (my fault due to excessive over-clocking which caused several reboots on one machine) I could get roughly the same hashrate as claymore's (10.5) with lower overclocking and more undervoltage. (With Gtx 1050 Ti i can get the max stable hashrate of 14.25 Mhs with only 52.50 power draw)

If I may just add few things:

1. I second the suggestion to be consistent with output logging (many of us parse the logs to enable their personal watchdogs)
2. I'd like to see in output the time of the record filled with milliseconds eg. hh:mm:ss.nnn

Bottom line : thank you VERY MUCH !!!

**Comment by AndreaLanfranchi** [comment]
I have to record another beneficial improvement of rc9.
Self recovery of failed GPU

Take note on this log :

```
  ℹ  13:11:09|stratum    Accepted. 
  m  13:11:14|ethminer  Speed 196.46 Mh/s    gpu/0 32.85  gpu/1 32.77  gpu/2 32.45  gpu/3 32.77  gpu/4 32.85  gpu/5 32.77  [A2529+9:R0+0:F0] Time: 14:22
  ℹ  13:11:18|stratum   Received new job #03a2e332  seed: #51e74ac038c9e3e66658ef25c1591022  target: #0000000112e0be826d694b2e
  ℹ  13:11:22|stratum   Received new job #f9633797  seed: #51e74ac038c9e3e66658ef25c1591022  target: #0000000112e0be826d694b2e
  ℹ  13:11:24|stratum   Received new job #e1f58a9c  seed: #51e74ac038c9e3e66658ef25c1591022  target: #0000000112e0be826d694b2e
  m  13:11:24|ethminer  Speed 177.56 Mh/s    gpu/0 32.64  gpu/1 32.73  gpu/2 14.11  gpu/3 32.73  gpu/4 32.64  gpu/5 32.73  [A2529+9:R0+0:F0] Time: 14:22
  ℹ  13:11:28|cuda-4    Solution found; Submitted to eu1.ethermine.org
  ℹ  13:11:28|cuda-4    Nonce: 0x6a9e70a8b5467b3f
  ℹ  13:11:29|stratum    Accepted. 
  m  13:11:34|ethminer  Speed 163.54 Mh/s    gpu/0 32.71  gpu/1 32.71  gpu/2  0.00  gpu/3 32.71  gpu/4 32.71  gpu/5 32.71  [A2530+9:R0+0:F0] Time: 14:22
  
                                                                       GPU 2 FAILED 
                                          [ AFTER 20 SECONDS ... ]
  
  m  13:21:54|ethminer  Speed 164.11 Mh/s    gpu/0 32.79  gpu/1 32.79  gpu/2  0.00  gpu/3 32.79  gpu/4 32.87  gpu/5 32.87  [A2549+9:R0+0:F0] Time: 14:32
  m  13:22:04|ethminer  Speed 164.43 Mh/s    gpu/0 32.93  gpu/1 32.93  gpu/2  0.00  gpu/3 32.85  gpu/4 32.85  gpu/5 32.85  [A2549+9:R0+0:F0] Time: 14:33
  ℹ  13:22:06|cuda-0    Solution found; Submitted to eu1.ethermine.org
  ℹ  13:22:06|cuda-0    Nonce: 0x9c45731e2554c23e
  ℹ  13:22:06|stratum    Accepted. 
  ℹ  13:22:08|stratum   Received new job #20184617  seed: #51e74ac038c9e3e66658ef25c1591022  target: #0000000112e0be826d694b2e
  ℹ  13:22:11|stratum   Received new job #99f6ed1d  seed: #51e74ac038c9e3e66658ef25c1591022  target: #0000000112e0be826d694b2e
  m  13:22:14|ethminer  Speed 188.53 Mh/s    gpu/0 32.73  gpu/1 32.73  gpu/2 25.09  gpu/3 32.73  gpu/4 32.63  gpu/5 32.63  [A2550+9:R0+0:F0] Time: 14:33
  ℹ  13:22:19|cuda-0    Solution found; Submitted to eu1.ethermine.org
  ℹ  13:22:19|cuda-0    Nonce: 0x6e59762bcc28d5fb
  ℹ  13:22:19|stratum    Accepted. 
  ℹ  13:22:23|cuda-3    Solution found; Submitted to eu1.ethermine.org
  ℹ  13:22:23|cuda-3    Nonce: 0xb2264c1db26c2b64
  ℹ  13:22:23|stratum    Accepted. 
  m  13:22:24|ethminer  Speed 196.46 Mh/s    gpu/0 32.81  gpu/1 32.73  gpu/2 32.48  gpu/3 32.81  gpu/4 32.81  gpu/5 32.81  [A2552+9:R0+0:F0] Time: 14:33
  ℹ  13:22:26|stratum   Received new job #9b87fa42  seed: #51e74ac038c9e3e66658ef25c1591022  target: #0000000112e0be826d694b2e
  ℹ  13:22:30|cuda-1    Solution found; Submitted to eu1.ethermine.org
  ℹ  13:22:30|cuda-1    Nonce: 0x23c99b8a90698186
  ℹ  13:22:30|stratum    Accepted. 
  
                                                                       GPU 2 BACK AGAIN

```

**Comment by jean-m-cyr** [comment]
@AndreaLanfranchi 

> Self recovery of failed GPU

That, I can't explain!!!

**Comment by AndreaLanfranchi** [comment]
Well ... I report what I see.
But looking at jobs switching I see kick() somehow behaves like a restart.

Am I wrong ?

**Comment by jean-m-cyr** [comment]
@AndreaLanfranchi Keep reporting what you see. What I think should happen rarely matches what actually does! That's what makes it interesting.
